File: src/RentManager.Api/Models/Rent.cs
namespace RentManager.Api.Models;

public class Rent
{
    public int Id { get; set; }

    public int TenantId { get; set; }

    public Tenant Tenant { get; set; } = null!;

    public int Year { get; set; }

    public int Month { get; set; }

    // Frozen at generation time, so a later rent change does not rewrite
    // a month that already exists.
    public decimal AmountDue { get; set; }

    // Always re-derived as SUM(payments) by RentLedgerService, never
    // incremented in place.
    public decimal AmountPaid { get; set; }

    // Stored as a PostgreSQL date. No time, no timezone, no drift.
    public DateOnly DueDate { get; set; }

    public bool IsSettled { get; set; }

    public ICollection<Payment> Payments { get; set; } = new List<Payment>();
}

File: rent-manager-web/src/app/models/rent.ts
// Status comes from the server, never calculated in the browser.
export type RentStatus = 'Upcoming' | 'Pending' | 'Outstanding' | 'Paid';

export interface Payment {
  id: number;
  rentId: number;
  // Plain calendar date, e.g. "2026-08-27". No time, no timezone.
  paymentDate: string;
  amount: number;
  paymentMode: string;
  note: string | null;
}

export interface RentMonth {
  id: number;
  year: number;
  month: number;
  monthLabel: string;
  amountDue: number;
  amountPaid: number;
  remaining: number;
  dueDate: string;
  isSettled: boolean;
  status: RentStatus;
  timing: string;
  daysUntilDue: number;
  isDueSoon: boolean;
  // Whether this month can be collected right now.
  isPayable: boolean;
  payments: Payment[];
}

export interface CurrentMonth {
  rentId: number;
  year: number;
  month: number;
  monthLabel: string;
  amountDue: number;
  amountPaid: number;
  remaining: number;
  dueDate: string;
  status: RentStatus;
  timing: string;
  daysUntilDue: number;
  isDueSoon: boolean;
  isPayable: boolean;
}

// The next rent that exists but is not collectable yet.
export interface NextExpected {
  monthLabel: string;
  amountDue: number;
  dueDate: string;
  timing: string;
}

export interface TenantLedger {
  tenant: {
    id: number;
    name: string;
    mobileNumber: string;
    monthlyRent: number;
    rentDueDay: number;
    shopName: string | null;
  };
  today: string;
  currentMonth: CurrentMonth | null;
  previousOutstanding: number;
  // The single "Total to be Paid" figure.
  totalPayable: number;
  nextExpected: NextExpected | null;
  upcomingRentAmount: { amount: number; effectiveFrom: string } | null;
  history: RentMonth[];
}
